grype/matcher/apk/matcher.go

```diff
@@ -135,3 +135,81 @@ func vulnerabilitiesByID(vulns []vulnerability.Vulnerability) map[string][]vulne
 
 	return results
 }
+
+func (m *Matcher) findApkPackage(store vulnerability.Provider, d *distro.Distro, p pkg.Package) ([]match.Match, error) {
+	var matches = make([]match.Match, 0)
+
+	// find Alpine SecDB matches for the given package name and version
+	secDbMatches, err := common.FindMatchesByPackageDistro(store, d, p, m.Type())
+	if err != nil {
+		return nil, err
+	}
+
+	cpeMatches, err := m.cpeMatchesWithoutSecDbFixes(store, d, p)
+	if err != nil {
+		return nil, err
+	}
+
+	// keep all secdb matches, as this is an authoritative source
+	matches = append(matches, secDbMatches...)
+
+	// keep only unique CPE matches
+	matches = append(matches, deduplicateMatches(secDbMatches, cpeMatches)...)
+
+	return matches, nil
+}
+
+func (m *Matcher) matchBySourceIndirection(store vulnerability.Provider, d *distro.Distro, p pkg.Package) ([]match.Match, error) {
+	metadata, ok := p.Metadata.(pkg.ApkMetadata)
+	if !ok {
+		return nil, nil
+	}
+
+	// ignore packages without source indirection hints or where source name is identical to package name
+	if metadata.OriginPackage == "" || metadata.OriginPackage == p.Name {
+		return []match.Match{}, nil
+	}
+
+	// use source package name for exact package name matching
+	var indirectPackage pkg.Package
+
+	err := copier.Copy(&indirectPackage, p)
+	if err != nil {
+		return nil, fmt.Errorf("failed to copy package: %w", err)
+	}
+
+	// use the source package name
+	indirectPackage.Name = metadata.OriginPackage
+
+	//For each cpe, replace pkg name with origin and add to set
+	CPEStrings := strset.New()
+	for _, cpe := range indirectPackage.CPEs {
+		updatedCPEString := strings.Replace(cpe.BindToFmtString(), p.Name, indirectPackage.Name, -1)
+		CPEStrings.Add(updatedCPEString)
+	}
+
+	//With each entry in set, convert string to CPE and update indirectPackage CPEs
+	updatedCPEs := make([]syftPkg.CPE, 0)
+	for _, CPEString := range CPEStrings.List() {
+		updatedCPE, _ := syftPkg.NewCPE(CPEString)
+		updatedCPEs = append(updatedCPEs, updatedCPE)
+	}
+	indirectPackage.CPEs = updatedCPEs
```

Comment:
I'm wondering if this function could be improved by factoring out this section of code. As I'm reading this function, it seems like the function's task can be decomposed into other identifiable tasks, e.g.:

1. compute an "indirect package"
2. try to use that package to find matches
3. map these matches back to the original package as "indirect matches"

Is this a correct summary? And if it is, could we factor out step 1 into its own function?

Let me know if this would be easier to think through in a sync up!

Reply:
All makes sense! Moved building the indirect package into its own function

Reply:
Great improvement — one question here (I should've just put in this thread): https://github.com/anchore/grype/pull/407#discussion_r704735378